### PR TITLE
in_tail: avoid unnecessary open()s

### PR DESCRIPTION
**Pull request by dwoo4dwoo**
As discussed in https://groups.google.com/forum/#!topic/fluentd/BBRde4Lyl9E

### CODE REVIEW

**Comment by dwoo4dwoo** [comment]
The Travis CI build failures seem unrelated to my change ... (UdpInputTest EADDRINUSE)

**Comment by repeatedly** [comment]
Possible file rotation cases:

1. File is rotated before `stat = FileWrapper.stat(@path)`

No problem.

2. File is rotated ater `stat = FileWrapper.stat(@path)`

2a. `if @inode != inode || fsize < @fsize` is false

No problem. Next trigger should detect file rotation.

2b. `if @inode != inode || fsize < @fsize` is true

It seems no problem. `on_rotate` also checks stat info with previous file.
So this patch should not effect existing behaviour.

Does anyone have any concern for this patch?

**Comment by repeatedly** [comment]
There are no problem, I will merge this patch today.

File: lib/fluent/plugin/in_tail.rb [hunk 1]
```diff
@@ -593,13 +592,11 @@ def on_notify
           begin
             if @inode != inode || fsize < @fsize
               # rotated or truncated
+              io = FileWrapper.open(@path)
```
Comment (sonots) [inline lib/fluent/plugin/in_tail.rb]: When Errno::ENOENT occurs on `FileWrapper.stat(@path)`, it seems Errno::ENOENT occurs again at here. Do not you need to handle it?

Comment (repeatedly) [inline lib/fluent/plugin/in_tail.rb]: So adding `rescue nil` is better?

Comment (dwoo4dwoo) [inline lib/fluent/plugin/in_tail.rb]: @sonots : good catch - thanks!
@repeatedly : I'll update the pull request with your suggestion

**Comment by repeatedly** [comment]
@sonots Could you check again?

File: lib/fluent/plugin/in_tail.rb [hunk 2]
```diff
@@ -593,13 +592,11 @@ def on_notify
           begin
             if @inode != inode || fsize < @fsize
               # rotated or truncated
+              io = FileWrapper.open(@path) rescue nil
```
Comment (sonots) [inline lib/fluent/plugin/in_tail.rb]: Could you tell me when this io object will be closed? Or, not necessary to close? Why?

Comment (sonots) [inline lib/fluent/plugin/in_tail.rb]: I am not sure why there was

     io = nil

before

Comment (sonots) [inline lib/fluent/plugin/in_tail.rb]: You should ignore only Errno::ENOENT as repeatedly wrote.

Comment (dwoo4dwoo) [inline lib/fluent/plugin/in_tail.rb]: Hi @sonots ,

> Could you tell me when this io object will be closed?

The file descriptor is consumed here:

https://github.com/fluent/fluentd/blob/master/lib/fluent/plugin/in_tail.rb#L384

The first time through ... an io_handler is created here:

https://github.com/fluent/fluentd/blob/master/lib/fluent/plugin/in_tail.rb#L416

The file descriptor remains open until the next file rotation, removal, or fluentd shutdown.  The logic to deal with some of these situations begins here:

https://github.com/fluent/fluentd/blob/master/lib/fluent/plugin/in_tail.rb#L420

Other cases are handled here:

https://github.com/fluent/fluentd/blob/master/lib/fluent/plugin/in_tail.rb#L188

For example, shutdown calls stop_watchers().

> You should ignore only Errno::ENOENT as repeatedly wrote

Sorry, I thought @repeatedly was giving me a choice, and I picked the first option.  I'll update the pull request using the second option.  I'll also rebase against master.

David

**Comment by sonots** [comment]
LGTM

**Comment by repeatedly** [comment]
Thanks for the patch!

**Pull request merged**